### PR TITLE
mgr/orch: service_id can contain a '.' char (mds, nfs, iscsi)

### PR DESCRIPTION
**Pull request by mgfritch**
Match last occurrence of the hostname substr when parsing the daemon_id

* The service_id can contain a '.' char (mds, nfs, iscsi)
* rgw realm/zone could contain 'hostname'

Fixes: https://tracker.ceph.com/issues/45293
Fixes: https://tracker.ceph.com/issues/45294
Signed-off-by: Michael Fritch <mfritch@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mgfritch** [comment]
jenkins test make check

File: src/pybind/mgr/cephadm/tests/test_spec.py [hunk 1]
```diff
@@ -288,3 +305,148 @@ def test_rgw_service_name(spec: RGWSpec, dd: DaemonDescription, valid):
...
+        True
+    ),
+])
+def test_mds_service_name(spec: ServiceSpec, dd: DaemonDescription, valid):
+    if valid:
+        assert spec.service_name() == dd.service_name()
+    else:
+        with pytest.raises(OrchestratorError):
+            dd.service_name()
+
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/tests/test_spec.py:344]: The test function is the same for all test cases.  what about simply adding those to `test_rgw_service_name` while renaming it to `test_daemon_description_service_name`?

File: src/pybind/mgr/orchestrator/_interface.py [hunk 2]
```diff
@@ -1292,20 +1295,22 @@ def matches_service(self, service_name):
...
-            if self.hostname and self.hostname in self.daemon_id:
-                pre, post_ = self.daemon_id.split(self.hostname)
-                return pre[:-1]
-            else:
-                # daemon_id == "realm.zone.host.random"
-                v = self.daemon_id.split('.')
-                if len(v) == 4:
-                    return '.'.join(v[0:2])
-                # subcluster or fqdn? undecidable.
-                raise OrchestratorError(f"DaemonDescription: Cannot calculate service_id: {v}")
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/_interface.py:1305]: can we keep this special handing for RGWs?

Comment (mgfritch) [reply src/pybind/mgr/orchestrator/_interface.py:1305]: do we need to keep it around for some reason?

**Comment by mgfritch** [comment]
http://pulpito.ceph.com/mgfritch-2020-05-05_02:07:06-rados-wip-mgfritch-testing-2020-05-04-1622-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
> http://pulpito.ceph.com/mgfritch-2020-05-05_02:07:06-rados-wip-mgfritch-testing-2020-05-04-1622-distro-basic-smithi/

https://tracker.ceph.com/issues/45335

File: src/pybind/mgr/orchestrator/_interface.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/_interface.py:1305]: It's just we end up in an endless loop, if we mess up this method. I'm just a bit nervous.

**Pull request merged**